// @polkadot/ui-app/Status/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _types = require("@polkadot/types");

var _AddressMini = _interopRequireDefault(require("../AddressMini"));

var _Icon = _interopRequireDefault(require("../Icon"));

var _util = require("../util");

var _translate = _interopRequireDefault(require("../translate"));

// Copyright 2017-2019 @polkadot/ui-app authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
class Status extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    this.renderStatus = (_ref) => {
      let account = _ref.account,
          action = _ref.action,
          id = _ref.id,
          message = _ref.message,
          removeItem = _ref.removeItem,
          status = _ref.status;
      const addressRendered = account ? _react.default.createElement(_AddressMini.default, {
        value: account
      }) : undefined;
      return _react.default.createElement("div", {
        className: (0, _util.classes)('item', status),
        onClick: removeItem,
        key: id
      }, _react.default.createElement("div", {
        className: "wrapper"
      }, _react.default.createElement("div", {
        className: "container"
      }, _react.default.createElement("div", {
        className: "desc"
      }, _react.default.createElement("div", {
        className: "header"
      }, action), addressRendered, _react.default.createElement("div", {
        className: "status"
      }, message)), _react.default.createElement("div", {
        className: "short"
      }, _react.default.createElement(_Icon.default, {
        name: this.iconName(status)
      })))));
    };

    this.renderItem = (_ref2) => {
      let id = _ref2.id,
          extrinsic = _ref2.extrinsic,
          error = _ref2.error,
          removeItem = _ref2.removeItem,
          rpc = _ref2.rpc,
          status = _ref2.status;
      let method = rpc.method,
          section = rpc.section;

      if (extrinsic) {
        const found = _types.Method.findFunction(extrinsic.callIndex);

        if (found.section !== 'unknown') {
          method = found.method;
          section = found.section;
        }
      }

      const icon = this.signerIconName(status);
      return _react.default.createElement("div", {
        className: (0, _util.classes)('item', status),
        onClick: removeItem,
        key: id
      }, _react.default.createElement("div", {
        className: "wrapper"
      }, _react.default.createElement("div", {
        className: "container"
      }, _react.default.createElement("div", {
        className: "desc"
      }, _react.default.createElement("div", {
        className: "header"
      }, section, ".", method), _react.default.createElement("div", {
        className: "status"
      }, error ? error.message : status)), _react.default.createElement("div", {
        className: "short"
      }, _react.default.createElement(_Icon.default, {
        loading: icon === 'spinner',
        name: icon
      })))));
    };

    this.iconName = status => {
      switch (status) {
        case 'error':
          return 'ban';

        case 'event':
          return 'assistive listening devices';

        case 'received':
          return 'telegram plane';

        default:
          return 'check';
      }
    };

    this.signerIconName = status => {
      switch (status) {
        case 'cancelled':
          return 'ban';

        case 'completed':
        case 'finalized':
        case 'sent':
          return 'check';

        case 'dropped':
        case 'invalid':
        case 'usurped':
          return 'arrow down';

        case 'error':
          return 'warning sign';

        case 'queued':
          return 'random';

        default:
          return 'spinner';
      }
    };
  }

  render() {
    const _this$props = this.props,
          _this$props$stqueue = _this$props.stqueue,
          stqueue = _this$props$stqueue === void 0 ? [] : _this$props$stqueue,
          _this$props$txqueue = _this$props.txqueue,
          txqueue = _this$props$txqueue === void 0 ? [] : _this$props$txqueue;
    const allst = stqueue.filter((_ref3) => {
      let isCompleted = _ref3.isCompleted;
      return !isCompleted;
    });
    const alltx = txqueue.filter((_ref4) => {
      let status = _ref4.status;
      return !['completed', 'incomplete'].includes(status);
    });

    if (!allst.length && !alltx.length) {
      return null;
    }

    return _react.default.createElement("div", {
      className: "ui--Status"
    }, alltx.map(this.renderItem), allst.map(this.renderStatus));
  }

}

var _default = (0, _translate.default)(Status);

exports.default = _default;
