// @polkadot/ui-app/AddressMini.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _uiApp = require("@polkadot/ui-app");

var _util = require("./util");

var _Balance = _interopRequireDefault(require("./Balance"));

var _Bonded = _interopRequireDefault(require("./Bonded"));

var _IdentityIcon = _interopRequireDefault(require("./IdentityIcon"));

// Copyright 2017-2019 @polkadot/app-staking authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
class AddressMini extends _react.default.PureComponent {
  render() {
    const _this$props = this.props,
          children = _this$props.children,
          className = _this$props.className,
          _this$props$isPadded = _this$props.isPadded,
          isPadded = _this$props$isPadded === void 0 ? true : _this$props$isPadded,
          style = _this$props.style,
          value = _this$props.value;

    if (!value) {
      return null;
    }

    const address = value.toString();
    return _react.default.createElement("div", {
      className: (0, _util.classes)('ui--AddressMini', isPadded ? 'padded' : '', className),
      style: style
    }, _react.default.createElement("div", {
      className: "ui--AddressMini-info"
    }, _react.default.createElement(_IdentityIcon.default, {
      size: 24,
      value: address
    }), this.renderAddressOrName(address), children, this.renderOfflineStatus()), this.renderBalance(), this.renderBonded());
  }

  renderAddressOrName(address) {
    const _this$props2 = this.props,
          _this$props2$isShort = _this$props2.isShort,
          isShort = _this$props2$isShort === void 0 ? true : _this$props2$isShort,
          _this$props2$withAddr = _this$props2.withAddress,
          withAddress = _this$props2$withAddr === void 0 ? true : _this$props2$withAddr;

    if (!withAddress) {
      return null;
    }

    const name = (0, _util.getAddrName)(address);
    return _react.default.createElement("div", {
      className: "ui--AddressMini-address ".concat(name ? 'withName' : 'withAddr')
    }, name || (isShort ? (0, _util.toShortAddress)(address) : address));
  }

  renderBalance() {
    const _this$props3 = this.props,
          balance = _this$props3.balance,
          value = _this$props3.value,
          _this$props3$withBala = _this$props3.withBalance,
          withBalance = _this$props3$withBala === void 0 ? false : _this$props3$withBala;

    if (!withBalance || !value) {
      return null;
    }

    return _react.default.createElement(_Balance.default, {
      balance: balance,
      className: "ui--AddressSummary-balance",
      params: value
    });
  }

  renderBonded() {
    const _this$props4 = this.props,
          bonded = _this$props4.bonded,
          value = _this$props4.value,
          _this$props4$withBond = _this$props4.withBonded,
          withBonded = _this$props4$withBond === void 0 ? false : _this$props4$withBond;

    if (!withBonded || !value) {
      return null;
    }

    return _react.default.createElement(_Bonded.default, {
      bonded: bonded,
      className: "ui--AddressSummary-balance",
      label: "",
      params: value
    });
  }

  renderOfflineStatus() {
    const _this$props5 = this.props,
          value = _this$props5.value,
          offlineStatus = _this$props5.offlineStatus;

    if (!value || !offlineStatus) {
      return null;
    }

    return _react.default.createElement(_uiApp.RecentlyOffline, {
      accountId: value.toString(),
      offline: offlineStatus,
      tooltip: true,
      inline: true
    });
  }

}

exports.default = AddressMini;
