// @polkadot/ui-app/InputFile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDropzone = _interopRequireDefault(require("react-dropzone"));

var _util = require("./util");

var _Labelled = _interopRequireDefault(require("./Labelled"));

var _translate = _interopRequireDefault(require("./translate"));

// Copyright 2017-2019 @polkadot/ui-app authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
class InputFile extends _react.default.PureComponent {
  constructor() {
    super(...arguments);
    this.state = {};

    this.onDrop = files => {
      const onChange = this.props.onChange;
      files.forEach(file => {
        const reader = new FileReader();

        reader.onabort = () => {// ignore
        };

        reader.onerror = () => {// ignore
        }; // @ts-ignore ummm... events are not properly specified here?


        reader.onload = (_ref) => {
          let result = _ref.target.result;
          const data = new Uint8Array(result);
          const name = file.name;
          onChange && onChange(data, name);
          this.setState({
            file: {
              name,
              size: data.length
            }
          });
        };

        reader.readAsArrayBuffer(file);
      });
    };
  }

  render() {
    const _this$props = this.props,
          accept = _this$props.accept,
          className = _this$props.className,
          clearContent = _this$props.clearContent,
          help = _this$props.help,
          isDisabled = _this$props.isDisabled,
          _this$props$isError = _this$props.isError,
          isError = _this$props$isError === void 0 ? false : _this$props$isError,
          label = _this$props.label,
          placeholder = _this$props.placeholder,
          t = _this$props.t,
          withLabel = _this$props.withLabel;
    const file = this.state.file;
    return _react.default.createElement(_Labelled.default, {
      help: help,
      label: label,
      withLabel: withLabel
    }, _react.default.createElement(_reactDropzone.default, {
      accept: accept,
      className: (0, _util.classes)('ui--InputFile', isError ? 'error' : '', className),
      disabled: isDisabled,
      multiple: false,
      onDrop: this.onDrop
    }, _react.default.createElement("div", {
      className: "label"
    }, !file || clearContent ? placeholder || t('drag and drop the file here') : placeholder || t('{{name}} ({{size}} bytes)', {
      replace: file
    }))));
  }

}

var _default = (0, _translate.default)(InputFile);

exports.default = _default;
