// @polkadot/ui-api/with/callDiv.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = withCallDiv;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _call = _interopRequireDefault(require("./call"));

// Copyright 2017-2019 @polkadot/ui-api authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
function withCallDiv(endpoint) {
  let options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return function (render) {
    let defaultProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    class Inner extends _react.default.PureComponent {
      render() {
        const _this$props = this.props,
              callResult = _this$props.callResult,
              callUpdated = _this$props.callUpdated,
              children = _this$props.children,
              _this$props$className = _this$props.className,
              className = _this$props$className === void 0 ? defaultProps.className : _this$props$className,
              _this$props$label = _this$props.label,
              label = _this$props$label === void 0 ? '' : _this$props$label,
              style = _this$props.style;
        return _react.default.createElement("div", (0, _extends2.default)({}, defaultProps, {
          className: [className, callUpdated ? 'rx--updated' : undefined].join(' '),
          style: style
        }), label, render(callResult), children);
      }

    }

    return (0, _call.default)(endpoint, (0, _objectSpread2.default)({}, options, {
      propName: 'callResult'
    }))(Inner);
  };
}
