// @polkadot/ui-api/with/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "withApi", {
  enumerable: true,
  get: function get() {
    return _api.default;
  }
});
Object.defineProperty(exports, "withCall", {
  enumerable: true,
  get: function get() {
    return _call.default;
  }
});
Object.defineProperty(exports, "withCalls", {
  enumerable: true,
  get: function get() {
    return _calls.default;
  }
});
Object.defineProperty(exports, "withCallDiv", {
  enumerable: true,
  get: function get() {
    return _callDiv.default;
  }
});
Object.defineProperty(exports, "withMulti", {
  enumerable: true,
  get: function get() {
    return _multi.default;
  }
});
Object.defineProperty(exports, "withObservable", {
  enumerable: true,
  get: function get() {
    return _observable.default;
  }
});

var _api = _interopRequireDefault(require("./api"));

var _call = _interopRequireDefault(require("./call"));

var _calls = _interopRequireDefault(require("./calls"));

var _callDiv = _interopRequireDefault(require("./callDiv"));

var _multi = _interopRequireDefault(require("./multi"));

var _observable = _interopRequireDefault(require("./observable"));
